File: array/9-backspace-string-compare/backspace-string-compare.py
class Solution1:
    """解法一: 堆栈
    """

    def backspaceCompare(self, s: str, t: str) -> bool:
        def process_string(str):
            stack = []
            for char in str:
                if char != '#':
                    stack.append(char)
                elif stack:
                    stack.pop()
            return ''.join(stack)

        return process_string(s) == process_string(t)


class Solution2:
    """解法二: 快慢指针
    """

    def backspaceCompare(self, s: str, t: str) -> bool:
        s, t = list(s), list(t)
        s_slow = t_slow = 0
        
        for fast, x in enumerate(s):
            if x == '#':
                if s_slow != 0:
                    s_slow -= 1
                continue
            if s_slow != fast: s[s_slow] = s[fast]
            s_slow += 1
        
        for fast, x in enumerate(t):
            if x == '#':
                if t_slow != 0:
                    t_slow -= 1
                continue
            if t_slow != fast: t[t_slow] = t[fast]
            t_slow += 1

        return s[:s_slow] == t[:t_slow]

File: array/9-backspace-string-compare/backspace-string-compare.go
package main

type Solution1 struct{}

func (s1 Solution1) processString(str string) string {
	stack := make([]rune, 0)
	for _, char := range str {
		if char != '#' {
			stack = append(stack, char)
		} else if len(stack) > 0 {
			stack = stack[:len(stack)-1]
		}
	}
	return string(stack)
}

func (s1 Solution1) backspaceCompare(s string, t string) bool {
	return s1.processString(s) == s1.processString(t)
}

type Solution2 struct{}

func (s2 Solution2) processString(str []rune) []rune {
	slow := 0
	for fast := range str {
		if str[fast] == '#' {
			if slow != 0 {
				slow--
			}
			continue
		}
		if slow != fast {
			str[slow] = str[fast]
		}
		slow++
	}
	return str[:slow]
}

func (s2 Solution2) backspaceCompare(s string, t string) bool {
	sRunes, tRunes := []rune(s), []rune(t)
	return string(s2.processString(sRunes)) == string(s2.processString(tRunes))
}
